clients/vault/src/horizon/client.rs

```diff
@@ -0,0 +1,279 @@
+			.await
+			.map_err(|_| Error::HttpFetchingError)
+	}
+}
+
+struct HorizonFetcher<C: HorizonClient> {
+	client: C,
+	vault_secret_key: String,
+	last_tx_id: Option<Vec<u8>>,
+}
+
+impl<C: HorizonClient> HorizonFetcher<C> {
+	pub fn new(client: C, vault_secret_key: String) -> Self {
+		Self { client, vault_secret_key, last_tx_id: None }
+	}
+
+	/// Fetch recent transactions from remote and deserialize to HorizonResponse
+	/// Since the limit in the request url is set to one it will always fetch just one
+	async fn fetch_latest_txs(
+		&self,
+		is_public_network: bool,
+		next_page: PagingToken,
+	) -> Result<HorizonTransactionsResponse, Error> {
+		let vault_keypair: SecretKey = SecretKey::from_encoding(&self.vault_secret_key).unwrap();
+		let vault_address = vault_keypair.get_public();
+
+		let mut request_url = String::from(horizon_url(is_public_network)) +
+			"/accounts/" + str::from_utf8(vault_address.to_encoding().as_slice())
+			.map_err(|_| Error::HttpFetchingError)? +
+			"/transactions?";
+
+		if !next_page.is_empty() {
+			request_url.push_str(&format!("cursor={}&", next_page));
+		}
+		request_url += "order=desc&limit=20";
+
+		tracing::info!("request url: {:?}", request_url);
+
+		self.client.get_transactions(request_url.as_str()).await
+	}
+
+	fn is_unhandled_transaction(&mut self, tx: &Transaction) -> bool {
+		const UP_TO_DATE: () = ();
+		let latest_tx_id_utf8 = &tx.id;
+
+		let prev_tx_id = &self.last_tx_id;
+		let initial = !matches!(prev_tx_id, Some(_));
+
+		let result = match prev_tx_id {
+			Some(prev_tx_id) =>
+				if prev_tx_id == latest_tx_id_utf8 {
+					Err(UP_TO_DATE)
+				} else {
+					Ok(latest_tx_id_utf8.clone())
+				},
+			None => Ok(latest_tx_id_utf8.clone()),
+		};
+
+		match result {
+			Ok(latest_tx_id) => {
+				self.last_tx_id = Some(latest_tx_id.clone());
+				if !initial {
+					tracing::debug!(
+						"Found new transaction from Horizon (id {:#?}) to process...",
+						str::from_utf8(&latest_tx_id).unwrap()
+					);
+
+					true
+				} else {
+					tracing::info!("Initial transaction handled");
+					false
+				}
+			},
+			Err(UP_TO_DATE) => {
+				tracing::info!("Already up to date");
+				false
+			},
+		}
+	}
+
+	async fn fetch_horizon_and_process_new_transactions(
+		&mut self,
+		issue_set: Arc<Mutex<IssueRequests>>,
+		handler: Arc<Mutex<ScpMessageHandler>>,
+		next_page: PagingToken,
+	) -> Result<PagingToken, Error> {
+		let handler = handler.lock().await;
+
+		let res = self.fetch_latest_txs(handler.is_public_network, next_page).await;
+		let transactions = match res {
+			Ok(txs) => txs._embedded.records,
+			Err(e) => {
+				tracing::warn!("Failed to fetch transactions: {:?}", e);
+				return Ok(PagingToken::new())
+			},
+		};
+
+		let mut paging_token = PagingToken::new();
+		for transaction in transactions {
+			// update the paging_token
+			if let Some(page) = transaction.paging_token() {
+				paging_token = page;
+			} else {
+				paging_token = PagingToken::new();
+			}
```

Comment:
Is this needed? I would assume every transaction has a paging token, or is that not the case?
Since you instantiate the paging_token outside the loop anyways, I guess it is safe to remove the else branch.

Reply:
Also, I think I would prefer if we can find the highest paging_token in all transactions and choose it to be the new one explicitly. 
Doing it inside this loop might work but then we rely on the transactions to be in a particular order (ie, the latest transaction has to be the last one in the vector), otherwise the paging token is not set to the latest and we might fetch and process duplicate transactions eventually. 
Since the paging_token should always increase over time it should be safe to assume that the highest is the latest.